game-core/src/main/java/games/strategy/engine/data/GameParser.java

```diff
@@ -1204,72 +1256,124 @@ private void parseRepairFrontierRules(final List<Element> elements, final Repair
-      throws GameParseException {
-    final List<Tuple<String, String>> options = new ArrayList<>();
-    for (final Element current : values) {
+  private List<Tuple<String, String>> setOptions(final IAttachment attachment, final List<Element> options,
+      final Map<String, String> foreach, final Map<String, List<String>> variables) throws GameParseException {
+    final List<Tuple<String, String>> results = new ArrayList<>();
+    for (final Element option : options) {
       // decapitalize the property name for backwards compatibility
-      final String name = decapitalize(current.getAttribute("name"));
+      final String name = decapitalize(option.getAttribute("name"));
       if (name.isEmpty()) {
-        throw newGameParseException("Option name with zero length");
+        throw newGameParseException("Option name with zero length for attachment: " + attachment.getName());
+      }
+      final String value = option.getAttribute("value");
+      if (containsEmptyForeachVariable(value, foreach)) {
+        continue; // Skip adding option if contains empty foreach variable
+      }
+      final String valueWithForeach = replaceForeachVariables(value, foreach);
+      final String count = option.getAttribute("count");
+      String optionValues = (!count.isEmpty() ? count + ":" : "") + valueWithForeach;
+      if (!variables.isEmpty()) {
+        final List<String> listWithVariables = new ArrayList<>();
+        for (final String s : Splitter.on(':').split(optionValues)) {
+          listWithVariables.add(variables.containsKey(s) ? Joiner.on(":").join(variables.get(s)) : s);
+        }
+        optionValues = Joiner.on(":").join(listWithVariables);
       }
-
-      final String value = current.getAttribute("value");
-      final String count = current.getAttribute("count");
-      final String itemValues = (count.length() > 0 ? count + ":" : "") + value;
       try {
         attachment.getProperty(name)
             .orElseThrow(() -> newGameParseException(String.format(
                 "Missing property definition for option '%s' in attachment '%s'",
                 name, attachment.getName())))
-            .setValue(itemValues);
+            .setValue(optionValues);
       } catch (final GameParseException e) {
         throw e;
       } catch (final Exception e) {
-        throw newGameParseException("Unexpected Exception while setting values for attachment" + attachment, e);
+        throw newGameParseException("Unexpected Exception while setting values for attachment: " + attachment, e);
       }
 
-      options.add(Tuple.of(name, itemValues));
+      results.add(Tuple.of(name, optionValues));
+    }
+    return results;
+  }
+
+  private String replaceForeachVariables(final String s, final Map<String, String> foreach) {
+    String result = s;
+    for (final Entry<String, String> entry : foreach.entrySet()) {
+      result = result.replace(entry.getKey(), entry.getValue());
```

Comment:
This could potentially consume a lot of memory, because for each iteration a new string is being created.
I'm not quite sure if it's indended for this function to replace variables recursively.
Example:
`$var -> $`, `$other -> test` and `$test -> abc,`
woudl cause `$var$other` to resolve to `abc`.

If this is undesired, I'd kindly ask you to have a look at https://github.com/triplea-game/triplea/blob/7a347695b0fe08975b438b498c92cf95eee59de4/game-core/src/main/java/org/triplea/util/LocalizeHtml.java#L51-L65
where I used RegEx to implement a custom replacement function a while back.

Reply:
This is meant to be recursive so you can combine unlimited depths of lists together.